content/blog/2020-10-tokio-0-3.md

```diff
@@ -0,0 +1,138 @@
+
+## Changes to IO traits
+
+Following the lead of @sfackler’s [RFC](https://github.com/rust-lang/rfcs/pull/2930), we changed the `AsyncRead` and `AsyncWrite` traits to support reads into uninitialized memory. This change *only* affects users who implement the `AsyncRead` or `AsyncWrite` traits or manually calling the `poll_*` methods. If the `AsyncReadExt` or `AsyncWriteExt` traits are used, no change is required.
+
+Below is a simplified version of the new `AsyncRead`:
+
+```rust
+pub trait AsyncRead {
+    fn poll_read(
+        self: Pin<&mut Self>, 
+        cx: &mut Context<'_>, 
+        buf: &mut ReadBuf<'_>
+    ) -> Poll<Result<()>>;
+}
+
+pub struct ReadBuf<'a> {
+    buf: &'a mut [MaybeUninit<u8>],
+    filled: usize,
+    initialized: usize,
+}
+
+impl<'a> ReadBuf<'a> {
+    // functions here, see RFC
+}
+```
+The `&mut [u8]` argument to the Tokio 0.2’s `poll_read` method has some unintended sharp edges: an implementer, not the end consumer, of the `AsyncRead` trait can *read* the data stored in the `&mut [u8]` slice. If a programmer creates a mutable slice of bytes (a `&mut [u8]`) that references uninitialized memory (e.g., the excess capacity in a vector), then *reading* the `&mut [u8]` data would cause undefined behavior. The new design mitigates this issue by providing a `ReadBuf` struct that tracks the uninitialized memory, removing the need to initialize memory.
+
+Additionally, the `poll_read_buf` and `poll_write_buf` methods have been removed from both traits. In practice, these methods were rarely implemented. Vectored operations will be implemented by types that support them.
+
+## Feature flag simplification
+
+We have reduced the amount of feature flags needed to use the core features of Tokio. The `dns`, `tcp`, `udp`,  and `uds` feature flags are collapsed into a single `net` feature flag. We have also combined `rt-core` and `rt-util` into a single `rt` feature flag. This `rt` feature flag now encompasses everything to execute futures except for the multi-threaded runtime which now lives under a `rt-multi-thread` feature flag (previously was called `rt-threaded`).
+
+- `dns`, tcp`, `udp` and `uds` -> `net`
+- `rt-util` and `rt-core` -> `rt`
+- `rt-threaded` -> `rt-multi-threaded`
+
+## Runtime and Builder refactor
+
+The Tokio team has removed  `threaded_scheduler` and `basic_scheduler` from the public API. Instead, the runtime `Builder` type introduces two constructors: one constructor for a multi-threaded scheduler and one for a current thread scheduler. These constructors provide a misuse-resistant form of building a runtime that *does not* require fiddling with feature flags and builder methods. Tokio will no longer choose the runtime variant based on feature flags when using `Runtime::new` but instead will default to the multi-threaded runtime and is only available under the `rt` feature flag. In addition, we’ve renamed the `core_threads` builder method to the more accurate `worker_threads`. 
```

Comment:
~~The Tokio team has removed  `threaded_scheduler` and `basic_scheduler` from the public API. Instead, ~~

This isn't important and can just be removed.

> These constructors provide a misuse-resistant form of building a runtime that *does not* require fiddling with feature flags and builder methods. 

I'm not following this.... maybe say something like:

"The new builder API provides consistency across feature flag permutations"